modules/uhttp/http.go

```diff
@@ -29,122 +29,111 @@ import (
 	"sync"
 	"time"
 
+	"go.uber.org/fx"
 	"go.uber.org/fx/auth"
-	"go.uber.org/fx/service"
+	"go.uber.org/fx/config"
 	"go.uber.org/fx/ulog"
 
 	"github.com/pkg/errors"
+	"github.com/uber-go/tally"
 	"go.uber.org/zap"
 )
 
 const (
-	// ContentType is the header key that contains the body type
-	ContentType = "Content-Type"
-	// ContentLength is the length of the HTTP body
-	ContentLength = "Content-Length"
-	// ContentTypeText is the plain content type
-	ContentTypeText = "text/plain"
-	// ContentTypeJSON is the JSON content type
-	ContentTypeJSON = "application/json"
-
-	// HTTP defaults
-	defaultTimeout = 60 * time.Second
-	defaultPort    = 3001
-
-	// Reporter timeout for tracking HTTP requests
-	defaultReportTimeout = 90 * time.Second
-
 	// default healthcheck endpoint
 	healthPath = "/health"
 
 	// default pprof endpoint
 	pprofPath = "/debug/pprof"
 )
 
-var _ service.Module = &Module{}
-
 // A Module is a module to handle HTTP requests
 type Module struct {
-	listener net.Listener
-	lock     sync.RWMutex
+	listener    net.Listener
+	lock        sync.RWMutex
+	l           *zap.Logger
+	handlerCtor fx.Component
 }
 
 // Config handles config for HTTP modules
 type Config struct {
-	Port    int           `yaml:"port"`
-	Timeout time.Duration `yaml:"timeout"`
+	Port    int           `yaml:"port" default:"3001"`
+	Timeout time.Duration `yaml:"timeout" default:"60s"`
 	Debug   bool          `yaml:"debug" default:"true"`
 }
 
-// GetHandlersFunc returns http handler created by caller
-type GetHandlersFunc func(service service.Host) http.Handler
-
 // New returns a new HTTP ModuleProvider.
-func New(handlerFunc GetHandlersFunc) service.ModuleProvider {
-	return service.ModuleProviderFromFunc("uhttp", func(host service.Host) (service.Module, error) {
-		handler := handlerFunc(host)
-		return newModule(host, handler)
-	})
-}
-
-func newModule(host service.Host, handler http.Handler) (*Module, error) {
-	// setup config defaults
-	cfg := Config{
-		Port:    defaultPort,
-		Timeout: defaultTimeout,
-	}
-	log := ulog.Logger(context.Background()).With(zap.String("module", host.ModuleName()))
-	if err := host.Config().Get("modules").Get(host.ModuleName()).Populate(&cfg); err != nil {
-		log.Error("Error loading http module configuration", zap.Error(err))
+func New(handlerCtor fx.Component) *Module {
```

Comment:
I would really prefer to have it as an abstract function, that returns modules, we can pass configs/loggers/etc directly to modules then.